core/src/main/java/org/jdbi/v3/core/internal/CollectionFactory.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jdbi.v3.core.internal;
+
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
+import java.util.Collections;
+import java.util.Set;
+import java.util.function.Supplier;
+import java.util.stream.Collector;
+import java.util.stream.Collectors;
+
+import org.jdbi.v3.core.internal.exceptions.Unchecked;
+
+@SuppressWarnings("rawtypes")
+public final class CollectionFactory {
+    private static final Supplier<Collector> SET_COLLECTOR;
+    static {
+        Supplier<Collector> factory = null;
+        try {
+            MethodHandle mh = MethodHandles.publicLookup()
+                    .findStatic(Collectors.class, "toUnmodifiableSet", MethodType.methodType(Collector.class));
+            factory = Unchecked.supplier(() -> (Collector) mh.invokeExact());
+        } catch (final ReflectiveOperationException e) {
+            // ignored
+            e.fillInStackTrace();
+        }
+        if (factory == null) {
```

Comment:
I think most of this method would be neater as a static factory method that returns the instance and sets it statically, i.e. `private static final Supplier<Collector> SET_COLLECTOR = getOptimalInstance();`. That would avoid this ugly temp var logic and conditional (the logic of which can go into the `catch` clause).

Reply:
```java
    private static Supplier<Collector> getOptimalInstance() {
        try {
            MethodHandle mh = MethodHandles.publicLookup().findStatic(Collectors.class, "toUnmodifiableSet", MethodType.methodType(Collector.class));
            return Unchecked.supplier(() -> (Collector) mh.invokeExact());
        } catch (final ReflectiveOperationException ignored) {
            return () -> Collectors.collectingAndThen(
                Collectors.toSet(),
                Collections::unmodifiableSet
            );
        }
    }
```